File: csharp/src/Google.Protobuf.Test/WellKnownTypes/AnyTest.cs
#region Copyright notice and license
// Protocol Buffers - Google's data interchange format
// Copyright 2015 Google Inc.  All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file or at
// https://developers.google.com/open-source/licenses/bsd
#endregion

using Google.Protobuf.Reflection;
using Google.Protobuf.TestProtos;
using NUnit.Framework;
using System.Linq;
using UnitTest.Issues.TestProtos;

namespace Google.Protobuf.WellKnownTypes
{
    public class AnyTest
    {
        [Test]
        public void Pack()
        {
            var message = SampleMessages.CreateFullTestAllTypes();
            var any = Any.Pack(message);
            Assert.AreEqual("type.googleapis.com/protobuf_unittest3.TestAllTypes", any.TypeUrl);
            Assert.AreEqual(message.CalculateSize(), any.Value.Length);
        }

        [Test]
        public void Pack_WithCustomPrefix()
        {
            var message = SampleMessages.CreateFullTestAllTypes();
            var any = Any.Pack(message, "foo.bar/baz");
            Assert.AreEqual("foo.bar/baz/protobuf_unittest3.TestAllTypes", any.TypeUrl);
            Assert.AreEqual(message.CalculateSize(), any.Value.Length);
        }

        [Test]
        public void Pack_WithCustomPrefixTrailingSlash()
        {
            var message = SampleMessages.CreateFullTestAllTypes();
            var any = Any.Pack(message, "foo.bar/baz/");
            Assert.AreEqual("foo.bar/baz/protobuf_unittest3.TestAllTypes", any.TypeUrl);
            Assert.AreEqual(message.CalculateSize(), any.Value.Length);
        }

        [Test]
        public void Unpack_WrongType()
        {
            var message = SampleMessages.CreateFullTestAllTypes();
            var any = Any.Pack(message);
            Assert.Throws<InvalidProtocolBufferException>(() => any.Unpack<TestOneof>());
        }

        [Test]
        public void Unpack_Success()
        {
            var message = SampleMessages.CreateFullTestAllTypes();
            var any = Any.Pack(message);
            var unpacked = any.Unpack<TestAllTypes>();
            Assert.AreEqual(message, unpacked);
        }

        [Test]
        public void Unpack_CustomPrefix_Success()
        {
            var message = SampleMessages.CreateFullTestAllTypes();
            var any = Any.Pack(message, "foo.bar/baz");
            var unpacked = any.Unpack<TestAllTypes>();
            Assert.AreEqual(message, unpacked);
        }

        [Test]
        public void TryUnpack_WrongType()
        {
            var message = SampleMessages.CreateFullTestAllTypes();
            var any = Any.Pack(message);
            Assert.False(any.TryUnpack(out TestOneof unpacked));
            Assert.Null(unpacked);
        }

        [Test]
        public void TryUnpack_RightType()
        {
            var message = SampleMessages.CreateFullTestAllTypes();
            var any = Any.Pack(message);
            Assert.IsTrue(any.TryUnpack(out TestAllTypes unpacked));
            Assert.AreEqual(message, unpacked);
        }

        [Test]
        public void ToString_WithValues()
        {
            var message = SampleMessages.CreateFullTestAllTypes();
            var any = Any.Pack(message);
            var text = any.ToString();
            Assert.That(text, Does.Contain("\"@value\": \"" + message.ToByteString().ToBase64() + "\""));
        }

        [Test]
        [TestCase("proto://foo.bar", "foo.bar")]
        [TestCase("/foo/bar/baz", "baz")]
        [TestCase("foobar", "")]
        public void GetTypeName(string typeUrl, string expectedTypeName)
        {
            Assert.AreEqual(expectedTypeName, Any.GetTypeName(typeUrl));
        }

        [Test]
        public void ToString_Empty()
        {
            var any = new Any();
            Assert.AreEqual("{ \"@type\": \"\", \"@value\": \"\" }", any.ToString());
        }

        [Test]
        public void ToString_MessageContainingAny()
        {
            var message = new TestWellKnownTypes { AnyField = new Any() };
            Assert.AreEqual("{ \"anyField\": { \"@type\": \"\", \"@value\": \"\" } }", message.ToString());
        }

        [Test]
        public void IsWrongType()
        {
            var any = Any.Pack(SampleMessages.CreateFullTestAllTypes());
            Assert.False(any.Is(TestOneof.Descriptor));
        }

        [Test]
        public void IsRightType()
        {
            var any = Any.Pack(SampleMessages.CreateFullTestAllTypes());
            Assert.True(any.Is(TestAllTypes.Descriptor));
        }

        [Test]
        public void Unpack_TypeRegistry()
        {
            var messages = new IMessage[]
            {
                SampleMessages.CreateFullTestAllTypes(),
                new TestWellKnownTypes { BoolField = true },
                new MoreString { Data = { "x" } },
                new MoreBytes { Data = ByteString.CopyFromUtf8("xyz") },
                new ReservedNames { Descriptor_ = 10 }
            };
            var anyMessages = messages.Select(Any.Pack);

            // The type registry handles the first four of the packed messages, but not the final one.
            var registry = TypeRegistry.FromFiles(
                UnittestWellKnownTypesReflection.Descriptor,
                UnittestProto3Reflection.Descriptor);
            var unpacked = anyMessages.Select(any => any.Unpack(registry)).ToList();
            var expected = (IMessage[]) messages.Clone();
            expected[4] = null;
            Assert.AreEqual(expected, unpacked);
        }
    }
}

File: python/google/protobuf/internal/any_test.py
# -*- coding: utf-8 -*-
# Protocol Buffers - Google's data interchange format
# Copyright 2008 Google Inc.  All rights reserved.
#
# Use of this source code is governed by a BSD-style
# license that can be found in the LICENSE file or at
# https://developers.google.com/open-source/licenses/bsd

"""Tests proto Any APIs."""

import unittest

from google.protobuf import any as proto_any

from google.protobuf import any_pb2
from google.protobuf import unittest_pb2


class AnyTest(unittest.TestCase):

  def test_pack_unpack(self):
    all_types = unittest_pb2.TestAllTypes()
    any_msg = proto_any.pack(all_types)
    all_descriptor = all_types.DESCRIPTOR
    self.assertEqual(
        any_msg.type_url, 'type.googleapis.com/%s' % all_descriptor.full_name
    )

    # Any can be successfully unpacked to the correct message type.
    unpacked_message = unittest_pb2.TestAllTypes()
    self.assertTrue(proto_any.unpack(any_msg, unpacked_message))

    proto_any.unpack_as(any_msg, unittest_pb2.TestAllTypes)

    # Any can't be unpacked to an incorrect message type.
    self.assertFalse(
        proto_any.unpack(any_msg, unittest_pb2.TestAllTypes.NestedMessage())
    )

    with self.assertRaises(TypeError) as catcher:
      proto_any.unpack_as(any_msg, unittest_pb2.TestAllTypes.NestedMessage)
    self.assertIn('Attempted to unpack', catcher.exception.args[0])

  def test_type_name(self):
    all_types = unittest_pb2.TestAllTypes()
    any_msg = proto_any.pack(all_types)
    self.assertEqual(
        proto_any.type_name(any_msg), 'proto2_unittest.TestAllTypes'
    )

  def test_is_type(self):
    all_types = unittest_pb2.TestAllTypes()
    any_msg = proto_any.pack(all_types)
    all_descriptor = all_types.DESCRIPTOR
    self.assertTrue(proto_any.is_type(any_msg, all_descriptor))

    empty_any = any_pb2.Any()
    self.assertFalse(proto_any.is_type(empty_any, all_descriptor))


if __name__ == '__main__':
  unittest.main()
